crates/generator/src/generator.rs

```diff
@@ -179,16 +240,45 @@ impl Generator {
             let mut machine = TraceMachine::new(default_machine);
 
             machine.load_program(&backend.generator, &[])?;
-            match machine.run() {
+            let maybe_ok = machine.run();
+            let execution_cycles = machine.machine.cycles();
+            drop(machine);
+
+            // Subtract tx execution cycles.
+            if let Some(cycles_pool) = &mut cycles_pool {
+                if cycles_pool.checked_sub_cycles(execution_cycles).is_none() {
```

Comment:
We should use the name `consume_cycles`, it is hard to see this function had side-effect when read the code.